src/formats/MMTF.cpp

```diff
@@ -301,6 +305,115 @@ void MMTFFormat::read(Frame& frame) {
+            for (auto id : transform.chainIndexList) {
+                chains_to_transform.insert(static_cast<double>(id));
+            }
+
+            auto& ncs = transform.matrix;
+
+            auto rotation = Matrix3D(ncs[0], ncs[4], ncs[8],
+                                     ncs[1], ncs[5], ncs[9],
+                                     ncs[2], ncs[6], ncs[10]);
+            auto translation = Vector3D(ncs[3], ncs[7], ncs[11]);
+
+            if (rotation == Matrix3D::unit() && translation == Vector3D()) {
+                continue;
+            }
+
+            rotation = rotation.invert();
+
+            // See comment below
+            // std::unordered_map<size_t, size_t> old_to_sym;
+
+            std::vector<Residue> residues_to_add;
+            for (const auto& residue : frame.topology().residues()) {
+
+                auto asmbl = residue.get("assembly");
+                
+                if (!asmbl || asmbl->as_string() != "bio" + assembly.name) {
+                    continue;
+                }
+
+                auto chainindex = residue.get("chainindex");
+                if (!chainindex || chains_to_transform.count(chainindex->as_double()) == 0) {
+                    continue;
+                }
+
+                auto new_residue = Residue(residue.name(), *residue.id());
+
+                new_residue.set("assembly", *asmbl);
+                new_residue.set("composition_type", *residue.get("composition_type"));
+                new_residue.set("is_standard_pdb", *residue.get("is_standard_pdb"));
+                new_residue.set("chainid", *residue.get("chainname"));
+
+                auto chainname = residue.get("chainname");
+                if (chainname) {
+                    new_residue.set("chainname", *chainname);
+                }
+
+                auto ss = residue.get("secondary_structure");
+                if (ss) {
+                    new_residue.set("secondary_structure", *ss);
+                }
+
+                for (auto atom_id : residue) {
+
+                    if (atom_id >= original_size) {
+                        continue;
+                    }
+
+                    auto atom_copy = frame[atom_id];
+                    auto new_position = rotation * frame.positions()[atom_id] + translation;
+
+                    frame.add_atom(std::move(atom_copy), std::move(new_position));
+                    new_residue.add_atom(frame.size() - 1);
+                    // old_to_sym.insert({ atom_id, frame.size() - 1 });
+                }
+
+                residues_to_add.emplace_back(new_residue);
+            }
+
+            for (auto&& residue : residues_to_add) {
+                frame.add_residue(std::move(residue));
+            }
+            
+            /*
+
+            The code below is way too slow, even in release mode. It is probably
+            due to the ordered nature of the bonds array. Keeping the first, but
+            not second loop causes the slow down.
```

Comment:
so what you want here is the list of bonds containing a given atom? How about keeping such list inside the MMTF parser itself, and only inserting bonds in the frame when they have all been found, i.e. at the end of `read`?

Reply:
This worked!